core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala

```diff
@@ -0,0 +1,50 @@
+package aima.core.search.adversarial
+
+import scala.annotation.tailrec
+
+/**
+  * @author Aditya Lahiri
+  * @author Shawn Garner
+  */
+object MinimaxDecision {
+  def minMaxDecision[PLAYER <: Player, STATE <: State, ACTION <: Action](
+      g: Game[PLAYER, STATE, ACTION],
+      noAction: ACTION
+  ): (STATE) => ACTION = { (state: STATE) =>
+    @tailrec def maxMinValue(Actions: List[ACTION]): ACTION = Actions match {
+      case Nil                   => noAction
+      case singularAction :: Nil => singularAction
+      case action1 :: action2 :: rest =>
+        maxMinValue(
+          (if (minValue(g.result(state, action1)) > minValue(g.result(state, action2))) action1
+           else action2) :: rest
+        )
+    }
+
+    @tailrec def minMaxValue(Actions: List[ACTION]): ACTION = Actions match {
+      case Nil                   => noAction
+      case singularAction :: Nil => singularAction
+      case action1 :: action2 :: rest =>
+        minMaxValue(
+          (if (maxValue(g.result(state, action1)) < maxValue(g.result(state, action2))) action1
+           else action2) :: rest
+        )
+    }
+
+    def maxValue(state: STATE): UtilityValue = {
+      if (g.isTerminalState(state))
+        g.getUtility(state)
+      else
+        minValue(g.result(state, maxMinValue(g.getActions(state))))
+    }
+
+    def minValue(state: STATE): UtilityValue = {
+      if (g.isTerminalState(state))
+        g.getUtility(state)
+      else
+        maxValue(g.result(state, minMaxValue(g.getActions(state))))
+    }
+
+    maxMinValue(g.getActions(g.initialState))
```

Comment:
@a1shadows I was looking at this when trying to implement alpha beta search and have a question.  Why are you passing in the `g.initialState` rather than `state` which is passed into the function?  this does not seem to match the algorithm.